rfcs/0004-checkpoints.md

```diff
@@ -0,0 +1,315 @@
+3. Update the manifest’s `wal_id_last_seen` to the latest WAL id found in the wal directory
+4. Write the manifest. If this fails with a CAS error, go back to 1.
+
+Then, the reader will restore data from the WAL. The reader will just replay the WAL data into an in-memory table. The size of the table will be bounded by `max_memtable_bytes`. When the table fills up, it’ll be added to a list and a new table will be initialized. The total memory used should be limited by backpressure applied by the writer. This could still be a lot of memory (e.g. up to 256MB in the default case). In the future we could have the reader read directly from the WAL SSTs, but I’m leaving that out of this proposal.
+
+##### Checkpoint Maintenance
+
+If the reader is opened without a checkpoint, it will periodically re-establish a checkpoint by re-running the initialization process described above, with the following modifications:
+- It will delete the checkpoint that it had previously created (when we support range scans, we’ll need to reference count the checkpoint).
+- It will purge any in-memory tables with content from WAL SSTs lower than the new `wal_id_last_compacted`.
+
+#### Garbage Collector
+
+The GC task will be modified to handle soft deletes and manage checkpoints/clones. The GC tasks's algorithm will now look like the following:
+1. Read the current manifest. If no manifest is found, exit.
+2. If `destroyed_at_s` is set, and the current time is after `destroyed_at_s` + `db_delete_grace`, and there are no active checkpoints, then:
+    1. If the db is a clone, update the source db's manifest by deleting the checkpoint.
+    2. Delete all object's under the db's path.
+    3. Exit.
+3. Garbage collect expired checkpoints
+    1. Find any checkpoints that have expired and remove them
+    2. Write the new manifest version with checkpoints deleted. If CAS fails, go back to 1 to reload checkpoints that may be refreshed or added.
+4. Delete garbage manifests. This includes any manifest that is older than `min_age` and is not referenced by a checkpoint.
+5. Read all manifests referenced by a checkpoint. Call this set M
+6. Clean up WAL SSTs.
+    1. Let C be the lowest id of all `wal_id_last_compacted` in M.
+    2. Delete all WAL SSTs with id lower than C and age older than `min_age`
+7. Clean up SSTs.
+    1. Let S be the set of all SSTs from all manifests in M.
+    2. Delete all SSTs not in M with age older than `min_age`
+8. Detach the clone if possible.
+    1. If the DB instance is a clone, and it's manifest and contained checkpoints no longer references any SSTs from its `source_checkpoint` (we can tell this if all SSTs are under the current db's path), then detach it:
+        1. Update the source db's manifest by removing the checkpoint.
+        2. Update the db's manifest by removing the `source_checkpoint` field.
```

Comment:
WDYT about adding `detach()` method/CLI command that forces a full compaction (or maybe copy all SSTs?). Not necessarily for this PR, but I think it should be possible. Would be a cool way to do backups and migrations.

Reply:
The idea makes total sense to me. I feel like a good starting point would be to just support some way to manually trigger a major compaction, and then the db should get detached automatically once that happens. I don't think this needs to be specified in this RFC. I'll file an issue to cover this use case.

Reply:
https://github.com/slatedb/slatedb/issues/288